File: src/carrot_switch/backup.py
"""Backup management for configuration and skills."""
import shutil
from datetime import datetime
from pathlib import Path

BACKUP_ROOT = Path.home() / "AppData" / "Roaming" / ".carrotswitch"


def _timestamp() -> str:
    return datetime.now().strftime("%Y%m%d_%H%M%S")


def backup_config(agent: str, config_path: Path) -> Path:
    """Backup a config file before modification."""
    ts = _timestamp()
    dest_dir = BACKUP_ROOT / "mcp" / agent
    dest_dir.mkdir(parents=True, exist_ok=True)
    dest = dest_dir / f"config_{ts}.jsonc"
    if config_path.exists():
        shutil.copy2(config_path, dest)
    return dest


def backup_skill(agent: str, skill_name: str) -> Path | None:
    """Backup a skill directory before modification/deletion."""
    from carrot_switch.skill import paths as skill_paths

    skill_dir = skill_paths.get_user_skills_dir(agent) / skill_name
    if not skill_dir.exists():
        return None

    ts = _timestamp()
    dest_dir = BACKUP_ROOT / "skill" / agent
    dest_dir.mkdir(parents=True, exist_ok=True)
    dest = dest_dir / f"{skill_name}_{ts}"
    shutil.copytree(skill_dir, dest)
    return dest

File: server/src/lib/backup.ts
import { join } from "path";
import { existsSync, copyFileSync, cpSync, mkdirSync } from "fs";
import { getUserSkillsDir } from "./skill/paths.js";
import { BACKUP_ROOT } from "./base.js";

function timestamp(): string {
  const now = new Date();
  const y = now.getFullYear();
  const m = String(now.getMonth() + 1).padStart(2, "0");
  const d = String(now.getDate()).padStart(2, "0");
  const h = String(now.getHours()).padStart(2, "0");
  const min = String(now.getMinutes()).padStart(2, "0");
  const s = String(now.getSeconds()).padStart(2, "0");
  return `${y}${m}${d}_${h}${min}${s}`;
}

export function backupConfig(agent: string, configPath: string): string {
  const ts = timestamp();
  const destDir = join(BACKUP_ROOT, "mcp", agent);
  mkdirSync(destDir, { recursive: true });
  const dest = join(destDir, `config_${ts}.jsonc`);
  if (existsSync(configPath)) {
    copyFileSync(configPath, dest);
  }
  return dest;
}

export function backupSkill(agent: string, skillName: string): string | null {
  const skillDir = join(getUserSkillsDir(agent), skillName);
  if (!existsSync(skillDir)) return null;

  const ts = timestamp();
  const destDir = join(BACKUP_ROOT, "skill", agent);
  mkdirSync(destDir, { recursive: true });
  const dest = join(destDir, `${skillName}_${ts}`);
  cpSync(skillDir, dest, { recursive: true });
  return dest;
}
